Fix EZP-25642: Change "Send to Trash Button" to "Delete" button for users

crevillo opened the pull request:
Fixes: https://jira.ez.no/browse/EZP-25642

> Status: work-in-progress

I will need some guide to finish this properly. This tries to delete content insteand of sending content to trash when the content is user type. i have some questions about the code i've written specially for the the testing part. 

Is that the right way to check this? I mean, i'm duplicating lot of code there. Is there any way to extend tests here like we do with phpunit? 

Second. Do we need another icon for deleting or the trash one is enough?

##### tasks
- [x] Add delete or sendToTrash Button depending on content Type
- [x] Create content delete action
- [ ] Add icon for delete action

##### tests
manual and unit tests > in progress
 

ping @bdunogier @dpobel @StephaneDiot @yannickroger

crevillo commented:
added a new test to check delete action when the action bar is for a user. Let's see what travis has to say now...

bdunogier commented:
Also fails on master, with the same error. It needs to be fixed (remove memlimit), but it's not because of your code.

bdunogier commented:
Could you please edit the PR's title ? Work-in-progress doesn't have to be in the title, it is enough to add it in the description's head.

crevillo commented:
isn't it in the description head now?

dpobel commented on Tests/js/models/assets/ez-contentmodel-tests.js:
```diff
@@ -639,6 +639,72 @@ YUI.add('ez-contentmodel-tests', function (Y) {
+
+            Y.Mock.expect(this.contentService, {
+                method: 'deleteContent',
+                args: [this.contentId, Y.Mock.Value.Function],
```
you don't need 2 tests for this method. You only have to check that the delete `callback` is passed to the content service `deleteContent` method. you can write something like:

```js
var options = {api: this.capi},
    callback = function () {},
    content = this.model;

Y.Mock.expect(this.contentService, {
    method: 'deleteContent',
    args: [this.contentId, callback],
});
content.delete(options, callback);
```

dpobel commented on Tests/js/views/assets/ez-actionbarview-tests.js:
```diff
@@ -87,8 +107,126 @@ YUI.add('ez-actionbarview-tests', function (Y) {
 
     Y.Test.Runner.setName("eZ Action Bar View tests");
-    Y.Test.Runner.add(viewTest);
+    Y.Test.Runner.add(viewTest, userViewTest);
```
`add` only accepts one parameter so the `userViewTest` is completely ignored, you should add `Y.Test.Runner.add(userViewTest);` instead

dpobel commented on Tests/js/views/assets/ez-actionbarview-tests.js:
```diff
@@ -87,8 +107,126 @@ YUI.add('ez-actionbarview-tests', function (Y) {
+
+        tearDown: function () {
+            delete Y.eZ.CreateContentActionView;
+            delete Y.eZ.TranslateActionView;
```
missing `delete Y.eZ.MoveContentActionView`

crevillo replied:
@dpobel: Ok. But this was a copy paste from the other test.should we fix it in the other too?

dpobel replied:
yes

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -96,6 +97,26 @@ YUI.add('ez-locationviewviewservice', function (Y) {
 
         /**
+         * `deleteAction` event handler,
+         * it asks confirmation to the user before delete the location.
```
 `it asks confirmation to the user before deleting the content item.`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -96,6 +97,26 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+         * @protected
+         * @param {Object} e event facade of the deleteAction event
+         */
+        _deleteContentConfirmBox: function (e) {
```
I would call it `_confirmDeleteContent`

crevillo replied:
@dpobel: so, should i strip the `ConfirmBox` part?

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -96,6 +97,26 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+            this.fire('confirmBoxOpen', {
+                config: {
+                    title: "Are you sure you want to delete this content?",
+                    confirmHandler: Y.bind(function () {
```
nitpick: since the `_delete` parameter is not used, you can even write:
```js
confirmHandler: Y.bind(this._delete, this),
```

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -140,6 +161,30 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+         * @method _delete
+         * @protected
+         */
+        _delete: function () {
```
I would call it `_deleteContent`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -140,6 +161,30 @@ YUI.add('ez-locationviewviewservice', function (Y) {
         },
 
         /**
+         * Delete the location, triggering loading parent location and notifications
```
`Delete the content item`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -140,6 +161,30 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+         */
+        _delete: function () {
+            var that = this,
+                location = this.get('location'),
```
`that` is not needed.
The `location` and `locationId` are not really needed as well, you can use the content id in the notification identifier.

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +231,52 @@ YUI.add('ez-locationviewviewservice', function (Y) {
         },
 
         /**
+         * Send to trash callback triggering notifications and making app to navigate to parent location
```
`Delete content item callback`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +231,52 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+         * @method _afterDeleteCallback
+         * @protected
+         * @param {eZ.Location} parentLocation the parent location to which app will navigate to
+         * @param {eZ.Content} content the content to be trashed
```
`@param {eZ.Content} content the content item that has been deleted`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +231,52 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+            }
+
+            this._notify(
+                '"' + contentName + '" deleted',
```
`'"' + contentName + '" was deleted',`

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -140,6 +161,30 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+                parentLocation = path[path.length - 1];
+
+            this._notify(
+                'Deleting "' + content.get('name'),
```
missing ` + '"'` at the end of the string

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +231,52 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+
+            if (error) {
+                this._notify(
+                    'An error occurred when deleting "' + contentName,
```
missing ` + '"'` at the end of the string

dpobel commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +231,52 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+             * @event sentToTrash
+             * @param {eZ.Location} location
+             */
+            this.fire('delete', {location: location});
```
that's a bit too generic. I would call it `deleteContent` and I would also add the `content` to the event parameter. (also the `@event` in the doc is wrong)

dpobel commented:
nice start @crevillo :)

> Is that the right way to check this? I mean, i'm duplicating lot of code there. Is there any way to extend tests here like we do with phpunit? 

I guess you are talking about `_isNotActionCreated` and `_isActionCreated` in action bar view tests ? In such case, you can write something like:

```js
var viewTest,
    // others vars...
    isNotActionCreated = function () {
        // code here
    };

viewTest = new Y.Test.Case({
   name: "eZ Action Bar View test",
   setUp: function() { /* code */ },

   _isNotActionCreated: isNotActionCreated,

   // test using _isNotActionCreated
});
```

> Second. Do we need another icon for deleting or the trash one is enough?

I guess we can keep the current one for now, that's really a detail.

More important:

* the code you added to `eZ.LocationViewViewService` should be tested
* The `deleteContent` event is actually useful to make sure the tree does not display the content that was deleted so you should add it in https://github.com/ezsystems/PlatformUIBundle/blob/master/Resources/public/js/apps/plugins/ez-updatetreeplugin.js#L26 (with the corresponding test)

crevillo commented:
Thanks @dpobel, for the double check i was talking about how to check an action bar and a specific bar for users. Wondering if one test could extend another or something.

Thanka for the feedback. I Will take care of it

crevillo commented:
> The deleteContent event is actually useful to make sure the tree does not display the content that was deleted so you should add it in

the name of the event should be deleteContent or delete*d* content? asking because in the already defined ones we have `movedContent` or `publishedDraft`

crevillo commented:
@dpobel: Applied your comments and squashed everything in one commit. Let me know if I forgot something.

crevillo commented:
Build failing, but again looks because composer problems. 
re-ping @dpobel @StephaneDiot @yannickroger

dpobel commented:
you can ignore those failures since they are completely unrelated to your change and @bdunogier fixed that lately anyway in master and 1.2 branches.
(I'm reviewing your changes)

dpobel commented on Tests/js/views/services/assets/ez-locationviewviewservice-tests.js:
```diff
@@ -1368,10 +1368,520 @@ YUI.add('ez-locationviewviewservice-tests', function (Y) {
+                returns: 'alan-shearer'
+            });
+
+            Mock.expect(this.location, {
```
that's not needed but I guess this is a left over of another test

dpobel replied:
this is configuring the Location mock to have a `deleteContent` method which does not exist in the original Location Model so this method never called anywhere. If you remove that, the test still passes.
Note: the `parentLocationId` is not retrieved anywhere, the view service already has the path to the Location being displayed, so the parent Location is taken from it and that's the point of the line `this.service.set('path', [{location: parentLocation}]);` a few lines above.

dpobel commented on Tests/js/views/services/assets/ez-locationviewviewservice-tests.js:
```diff
@@ -1368,10 +1368,520 @@ YUI.add('ez-locationviewviewservice-tests', function (Y) {
+                returns: 'alan-shearer'
+            });
+
+            Mock.expect(this.location, {
```
same as https://github.com/ezsystems/PlatformUIBundle/pull/537/files#r58863220

dpobel replied:
and same following tests

crevillo replied:
i think i've copied all over the place... but don't you need it to get the `parentLocationId` and navigate to it?

dpobel commented on Tests/js/views/services/assets/ez-locationviewviewservice-tests.js:
```diff
@@ -1368,10 +1368,520 @@ YUI.add('ez-locationviewviewservice-tests', function (Y) {
+            this.service.set('path', [{location: parentLocation}]);
+            this.service.set('content', content);
+
+            Mock.expect(content, {
```
this mock configuration and the one for `parentLocation` are the same and repeated in every test, you should put that in one or 2 dedicated methods.

dpobel commented:
besides the last comment, that's good to me.
+1
ping @yannickroger

crevillo commented:
@dpobel, if i remove 
```js
 Mock.expect(parentLocation, {
     method: 'get',
     args: ['id'],
     returns: 'alan-shearer'
});
```
i get following errors running the tests...
```
✖ [eZ Location View View Service tests]: Passed: 34 Failed: 3 Total: 37 (ignored 0) (0.527 seconds)
    Should notify user about success of deleting the content item
       Unexpected error: 'undefined' is not a function (evaluating 'parentLocation.get('id')')
    Should fire deletedContent event after deleting a content item
       Unexpected error: 'undefined' is not a function (evaluating 'parentLocation.get('id')')
    Should navigate the app to view of parent location
       Unexpected error: 'undefined' is not a function (evaluating 'parentLocation.get('id')')
```

dpobel commented:
it seems I was unclear, sorry. You should remove only the following mock configuration:
```js
            Mock.expect(this.location, {
                method: 'deleteContent',
                args: [Mock.Value.Object, Mock.Value.Function],
                run: function (options, callback){
                    callback(true);
                }
            });
```

the other comment is about putting the common mock configuration in a dedicated method to not repeat that all over the test.

yannickroger commented:
+1
Cool to see you on the JS side of the force :smile:

crevillo commented:
ok, i'll try. thanks.

crevillo commented:
thanks @yannickroger, i think is not bad for my first js code in almost 7 years :dancer:

StephaneDiot commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +226,50 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+            if (error) {
+                this._notify(
+                    'An error occurred when deleting "' + contentName + '"',
+                    'content-' + content.get('id'),
```
Shouldn't it be `'delete-' + content.get('id')` here. Or your 'started' notification will not be updated and disapear.

crevillo replied:
you're right @StephaneDiot. My fault. i'll add these and blank lines fixes to last changes proposed by @dpobel. 
For the icon part, yep, @dpobel suggested me to use the same icon as the send to trash for now. 
Thanks for your input!.

StephaneDiot commented on Resources/public/js/views/services/ez-locationviewviewservice.js:
```diff
@@ -186,6 +226,50 @@ YUI.add('ez-locationviewviewservice', function (Y) {
+
+            this._notify(
+                '"' + contentName + '" was deleted',
+                'content-' + content.get('id'),
```
same here

StephaneDiot commented on Tests/js/views/services/assets/ez-locationviewviewservice-tests.js:
```diff
@@ -1368,10 +1368,520 @@ YUI.add('ez-locationviewviewservice-tests', function (Y) {
+                    } else {
+                        Y.fail("Unexpected parameter for content mock");
+                    }
+
```
niptick: unnecessary empty line

StephaneDiot commented on Tests/js/views/services/assets/ez-locationviewviewservice-tests.js:
```diff
@@ -1368,10 +1368,520 @@ YUI.add('ez-locationviewviewservice-tests', function (Y) {
+                    } else {
+                        Y.fail("Unexpected parameter for content mock");
+                    }
+
```
same niptick

StephaneDiot commented:
besides last comments +1, and good job ! For the icon thing, I don't know if you had your answer but you should ask to @dpobel .

crevillo commented:
@dpobel @StephaneDiot @yannickroger , added last comments
@dpobel: i finally added a `_initializeMocks` method that always do almost the same except the `true` or `false` for the content delete part. that's why i added another argument `callbackArgument`
maybe naming of those needs to be improved... suggestions?
thanks!.

yannickroger commented:
Maybe `error` ? As if `true` it will error.

crevillo commented:
you mean the callbackArgument param?

dpobel commented:
looks good to me. +1

(TravisCI failure does not seem to be related to your change, I guess it's one of our random failure... I restarted the job)
